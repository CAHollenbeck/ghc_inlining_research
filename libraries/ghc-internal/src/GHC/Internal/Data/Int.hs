{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Internal.Data.Int
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  portable
--
-- Signed integer types
--
-----------------------------------------------------------------------------

module GHC.Internal.Data.Int
    (
     -- * Signed integer types
     Int,
     Int8, Int16, Int32, Int64,
     ) where

import GHC.Internal.Base ( Int )
import GHC.Internal.Int  ( Int8, Int16, Int32, Int64 )
