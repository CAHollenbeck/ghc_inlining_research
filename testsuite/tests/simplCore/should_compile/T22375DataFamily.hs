{-# LANGUAGE GHC2021, TypeFamilies, StandaloneKindSignatures #-}

module T22375DataFamily where

import Data.Kind

type X :: Type -> Type
data family X a
data instance X () = A | B | C | D | E
  deriving Eq

f :: X () -> Int -> Int
f x v
  | x == A = 1 + v
  | x == B = 2 + v
  | x == C = 3 + v
  | x == D = 4 + v
  | otherwise = 5 + v
