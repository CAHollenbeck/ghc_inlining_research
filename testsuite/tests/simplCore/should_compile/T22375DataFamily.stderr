
==================== Tidy Core ====================
Result size of Tidy Core
  = {terms: 86, types: 65, coercions: 15, joins: 0/0}

-- RHS size: {terms: 1, types: 0, coercions: 2, joins: 0/0}
T22375DataFamily.$WA [InlPrag=INLINE[final] CONLIKE] :: X ()
[GblId[DataConWrapper],
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)}]
T22375DataFamily.$WA
  = T22375DataFamily.A
    `cast` (Sym (T22375DataFamily.D:R:XUnit0[0])
            :: T22375DataFamily.R:XUnit ~R# X ())

-- RHS size: {terms: 1, types: 0, coercions: 2, joins: 0/0}
T22375DataFamily.$WB [InlPrag=INLINE[final] CONLIKE] :: X ()
[GblId[DataConWrapper],
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)}]
T22375DataFamily.$WB
  = T22375DataFamily.B
    `cast` (Sym (T22375DataFamily.D:R:XUnit0[0])
            :: T22375DataFamily.R:XUnit ~R# X ())

-- RHS size: {terms: 1, types: 0, coercions: 2, joins: 0/0}
T22375DataFamily.$WC [InlPrag=INLINE[final] CONLIKE] :: X ()
[GblId[DataConWrapper],
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)}]
T22375DataFamily.$WC
  = T22375DataFamily.C
    `cast` (Sym (T22375DataFamily.D:R:XUnit0[0])
            :: T22375DataFamily.R:XUnit ~R# X ())

-- RHS size: {terms: 1, types: 0, coercions: 2, joins: 0/0}
T22375DataFamily.$WD [InlPrag=INLINE[final] CONLIKE] :: X ()
[GblId[DataConWrapper],
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)}]
T22375DataFamily.$WD
  = T22375DataFamily.D
    `cast` (Sym (T22375DataFamily.D:R:XUnit0[0])
            :: T22375DataFamily.R:XUnit ~R# X ())

-- RHS size: {terms: 1, types: 0, coercions: 2, joins: 0/0}
T22375DataFamily.$WE [InlPrag=INLINE[final] CONLIKE] :: X ()
[GblId[DataConWrapper],
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)}]
T22375DataFamily.$WE
  = T22375DataFamily.E
    `cast` (Sym (T22375DataFamily.D:R:XUnit0[0])
            :: T22375DataFamily.R:XUnit ~R# X ())

-- RHS size: {terms: 14, types: 11, coercions: 2, joins: 0/0}
T22375DataFamily.$fEqX_$c== :: X () -> X () -> Bool
[GblId,
 Arity=2,
 Str=<SL><SL>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=True)}]
T22375DataFamily.$fEqX_$c==
  = \ (a :: X ()) (b :: X ()) ->
      case GHC.Prim.dataToTagLarge#
             @GHC.Types.Lifted
             @T22375DataFamily.R:XUnit
             (a
              `cast` (T22375DataFamily.D:R:XUnit0[0]
                      :: X () ~R# T22375DataFamily.R:XUnit))
      of a#
      { __DEFAULT ->
      case GHC.Prim.dataToTagLarge#
             @GHC.Types.Lifted
             @T22375DataFamily.R:XUnit
             (b
              `cast` (T22375DataFamily.D:R:XUnit0[0]
                      :: X () ~R# T22375DataFamily.R:XUnit))
      of b#
      { __DEFAULT ->
      GHC.Prim.tagToEnum# @Bool (GHC.Prim.==# a# b#)
      }
      }

-- RHS size: {terms: 18, types: 11, coercions: 2, joins: 0/0}
T22375DataFamily.$fEqX_$c/= [InlPrag=INLINE (sat-args=2)]
  :: X () -> X () -> Bool
[GblId,
 Arity=2,
 Str=<SL><SL>,
 Unf=Unf{Src=StableUser, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=False,boring_ok=False)}]
T22375DataFamily.$fEqX_$c/=
  = \ (eta :: X ()) (eta1 :: X ()) ->
      case GHC.Prim.dataToTagLarge#
             @GHC.Types.Lifted
             @T22375DataFamily.R:XUnit
             (eta
              `cast` (T22375DataFamily.D:R:XUnit0[0]
                      :: X () ~R# T22375DataFamily.R:XUnit))
      of a#
      { __DEFAULT ->
      case GHC.Prim.dataToTagLarge#
             @GHC.Types.Lifted
             @T22375DataFamily.R:XUnit
             (eta1
              `cast` (T22375DataFamily.D:R:XUnit0[0]
                      :: X () ~R# T22375DataFamily.R:XUnit))
      of b#
      { __DEFAULT ->
      case GHC.Prim.==# a# b# of {
        __DEFAULT -> GHC.Types.True;
        1# -> GHC.Types.False
      }
      }
      }

-- RHS size: {terms: 3, types: 2, coercions: 0, joins: 0/0}
T22375DataFamily.$fEqX [InlPrag=CONLIKE] :: Eq (X ())
[GblId[DFunId],
 Unf=DFun: \ ->
       GHC.Classes.C:Eq TYPE: X ()
                        T22375DataFamily.$fEqX_$c==
                        T22375DataFamily.$fEqX_$c/=]
T22375DataFamily.$fEqX
  = GHC.Classes.C:Eq
      @(X ()) T22375DataFamily.$fEqX_$c== T22375DataFamily.$fEqX_$c/=

-- RHS size: {terms: 24, types: 4, coercions: 1, joins: 0/0}
T22375DataFamily.$wf [InlPrag=[2]]
  :: X () -> GHC.Prim.Int# -> GHC.Prim.Int#
[GblId[StrictWorker([!])],
 Arity=2,
 Str=<1L><L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [64 0] 55 0}]
T22375DataFamily.$wf
  = \ (x :: X ()) (ww :: GHC.Prim.Int#) ->
      case x
           `cast` (T22375DataFamily.D:R:XUnit0[0]
                   :: X () ~R# T22375DataFamily.R:XUnit)
      of {
        A -> GHC.Prim.+# 1# ww;
        B -> GHC.Prim.+# 2# ww;
        C -> GHC.Prim.+# 3# ww;
        D -> GHC.Prim.+# 4# ww;
        E -> GHC.Prim.+# 5# ww
      }

-- RHS size: {terms: 12, types: 6, coercions: 0, joins: 0/0}
f [InlPrag=[2]] :: X () -> Int -> Int
[GblId,
 Arity=2,
 Str=<1L><1!P(L)>,
 Cpr=1,
 Unf=Unf{Src=StableSystem, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)}]
f = \ (x :: X ()) (v :: Int) ->
      case v of { GHC.Types.I# ww ->
      case T22375DataFamily.$wf x ww of ww1 { __DEFAULT ->
      GHC.Types.I# ww1
      }
      }



